### PR TITLE
Signup: Remove hardcoded TLDs and load them from the API

### PR DESCRIPTION
**Pull request by aidvu**
- Remove hardcoded TLDs, as we have to maintain them whenever
we add a new TLD, while we already have the list on the backend
- Add a placeholder state while the product list is loading
- Create product list selector
- Refactor to use product list selector
- Instead of redirect, show a message when a TLD is not supported

Test:
- Visit http://calypso.localhost:3000/start/domain-first?new=testing-something-good-ol.li1ve
- See:
unsupported domain

---------

- Visit http://calypso.localhost:3000/start/domain-first?new=testing-something-good-ol.live
- See:
loading placeholders

- And after products being loaded:
original

---------

- Go through the flow and make sure everything is still working.

### CODE REVIEW

**Comment by aidvu** [comment]
Would be nice if @mikeshelton1503 can take a look. :)

File: client/signup/steps/site-or-domain/index.jsx [hunk 1]
```diff
@@ -18,21 +18,27 @@ import { getCurrentUserId } from 'state/current-user/selectors';
...
-			externalRedirect( '/domains' );
+	getDomainProductSlug( domain ) {
+		const domainParts = domain.split( '.' );
+		const tld = domainParts.slice( 1 ).join( '.' );
+
+		let productSlug = null;
+		if ( includes( [ 'com', 'net', 'org' ], tld ) ) {
+			productSlug = 'domain_reg';
+		} else {
+			productSlug = `dot${ tld }_domain`;
```
Comment (delputnam) [inline client/signup/steps/site-or-domain/index.jsx]: I wish we could use the `outer_slug` field to look up the `productSlug` instead of relying on this "magic" format. But I guess that won't work for 'com', 'net',  or 'org' TLDs anyway. :-/

Comment (aidvu) [inline client/signup/steps/site-or-domain/index.jsx]: Yeah, will think about it, might do it in another PR.

File: client/signup/steps/site-or-domain/index.jsx [hunk 2]
```diff
@@ -18,21 +18,27 @@ import { getCurrentUserId } from 'state/current-user/selectors';
...
 
 class SiteOrDomain extends Component {
-	componentWillMount() {
-		if ( ! this.getDomainName() ) {
-			// /domains domain search is an external application to calypso,
-			// therefor a full redirect required:
-			externalRedirect( '/domains' );
+	getDomainProductSlug( domain ) {
+		const domainParts = domain.split( '.' );
+		const tld = domainParts.slice( 1 ).join( '.' );
```
Comment (klimeryk) [inline client/signup/steps/site-or-domain/index.jsx]: We [have `getTld`](https://github.com/Automattic/wp-calypso/blob/master/client/lib/domains/index.js#L139) that should be helpful here :)

File: client/signup/steps/site-or-domain/index.jsx [hunk 3]
```diff
@@ -18,21 +18,27 @@ import { getCurrentUserId } from 'state/current-user/selectors';
...
+
+		let productSlug = null;
```
Comment (klimeryk) [inline client/signup/steps/site-or-domain/index.jsx]: You could just `return` ASAP and do away with `productSlug`:
```
if ( includes( [ 'com', 'net', 'org' ], tld ) ) {
	return 'domain_reg';
}

return `dot${ tld }_domain`;
```
(seems less complicated to me, but it's your call)

File: client/signup/steps/site-or-domain/index.jsx [hunk 4]
```diff
@@ -171,6 +182,32 @@ class SiteOrDomain extends Component {
...
+			const subHeaderText = translate(
+				'Please visit {{a}}wordpress.com/domains{{/a}} to search for another domain.',
+				{
+					components: {
+						a: <a href={ 'https://wordpress.com/domains' } />
+					}
+				}
+			);
+
+			return (
```
Comment (klimeryk) [inline client/signup/steps/site-or-domain/index.jsx]: Could you show something else than `StepWrapper`? It's very confusing - it says `Step 1 of 4`, but shows what's supposed to be an error message redirecting somewhere else. Maybe `EmptyContent` would be good instead?

File: client/state/products-list/selectors.js [hunk 5]
```diff
@@ -2,6 +2,10 @@ export function isProductsListFetching( state ) {
 	return state.productsList.isFetching;
 }
 
+export function getProductsList( state ) {
```
Comment (klimeryk) [inline client/state/products-list/selectors.js]: Nice 👍

File: client/signup/steps/site-or-domain/index.jsx [hunk 6]
```diff
@@ -171,6 +182,32 @@ class SiteOrDomain extends Component {
 	};
 
 	render() {
+		const { translate } = this.props;
+		const productsLoaded = ! isEmpty( this.props.productsList );
```
Comment (klimeryk) [inline client/signup/steps/site-or-domain/index.jsx]: You seem to be checking this a lot in this file - maybe it should a selector of its own? Then you'd have it in props and would not have to create it in four different places?

File: client/signup/steps/site-or-domain/index.jsx [hunk 7]
```diff
@@ -120,8 +128,11 @@ class SiteOrDomain extends Component {
 	}
 
 	renderScreen() {
+		const productsLoaded = ! isEmpty( this.props.productsList );
+
 		return (
 			<div>
+				{ ! productsLoaded && <QueryProductsList /> }
```
Comment (klimeryk) [inline client/signup/steps/site-or-domain/index.jsx]: Is this the correct pattern? Or should we always re-fetch? Seems like the latter is the more common pattern, though I might be wrong. We don't change the products list that often, but if we add a new TLD, we probably want to have it fetched here :)

**Comment by mikeshelton1503** [comment]
Looks good 👍  As discussed in Slack, using the step screen for a error message is not ideal, however given the rareness of that error I agree it's the best implementation for now.

I'm not sure about the placeholder here. It feels a bit useless. It technically follows our patterns but 3 big boxes don't really do much. Typically Calypso placeholders are used on individual elements within a card, not on the card itself. We also don't use them in other parts of NUX. If we're going to use them here I would suggest we show the card but use placeholders for the image, button, and copy, similar to this:

image

File: client/signup/steps/site-or-domain/index.jsx (hunk 4 above)
Comment (aidvu) [inline client/signup/steps/site-or-domain/index.jsx]: The step numbers are injected in the signup framework. Can't do it w/o major changes, or at least as far as I know. :)

File: client/signup/steps/site-or-domain/index.jsx (hunk 6 above)
Comment (aidvu) [inline client/signup/steps/site-or-domain/index.jsx]: Nice idea!

File: client/signup/steps/site-or-domain/index.jsx (hunk 7 above)
Comment (aidvu) [inline client/signup/steps/site-or-domain/index.jsx]: Yep, if it's already fetched it will be pulled from cache, and if it's the same the UI won't update so all cool. :)

File: client/components/data/query-products-list/index.jsx [hunk 8]
```diff
@@ -11,7 +11,7 @@ import { isProductsListFetching as isFetching } from 'state/products-list/select
 import { requestProductsList } from 'state/products-list/actions';
 
 class QueryProductsList extends Component {
-	componentWillMount() {
+	componentDidMount() {
```
Comment (aidvu) [inline client/components/data/query-products-list/index.jsx]: Now that I think about it, no idea. It made more sense to me at that point, but it's all the same.

**Comment by aidvu** [comment]
I see. Will update @mikeshelton1503!

**Comment by aidvu** [comment]
Updated. @klimeryk / @mikeshelton1503 one last check.

**Comment by mikeshelton1503** [comment]
Awesome, looks good. Design 👍

File: client/signup/steps/site-or-domain/style.scss [hunk 9]
```diff
@@ -45,6 +49,10 @@
 
 .site-or-domain__choice-text {
 	color: $gray-dark;
+
+	p {
+		min-height: 105px;
```
Comment (klimeryk) [inline client/signup/steps/site-or-domain/style.scss]: This rule along with https://github.com/Automattic/wp-calypso/issues/16319, causes this ugly result:
screen shot 2017-07-26 at 16 23 56

This is on newest Chromium and Safari on OSX, so I'd not consider it an edge case.

Comment (aidvu) [inline client/signup/steps/site-or-domain/style.scss]: :man_facepalming:

**Comment by klimeryk** [comment]
Just realized one more thing: this approach allows to potentially register any TLD that we offer or _used_ to offer:
screen shot 2017-07-26 at 16 29 57

Check http://calypso.localhost:3000/start/domain-first?new=lkf-ajsl-dkf-jak-lsj-dfk-la-jsdf.mx

**Comment by aidvu** [comment]
> Just realized one more thing: this approach allows to potentially register any TLD that we offer or used to offer.

Same as before. Fixing in a separate PR, good point. :)

**Comment by klimeryk** [comment]
Yup, that fixed the CSS issue - thanks! :shipit:
